cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/WindowState.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2017 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License"). See License in the project root for license information.
+ */
+
+package com.linkedin.cruisecontrol.monitor.sampling.aggregator;
+
+import com.linkedin.cruisecontrol.common.LongGenerationed;
+import com.linkedin.cruisecontrol.model.Entity;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static com.linkedin.cruisecontrol.monitor.sampling.aggregator.AggregationOptions.Granularity.ENTITY;
+
+
+/**
+ * A class that helps maintain the information of a window.
+ */
+public class WindowState<G, E extends Entity<G>> extends LongGenerationed {
+  private final Set<E> _validEntities;
+
+  public WindowState(long generation) {
+    super(generation);
+    _validEntities = new HashSet<>();
+  }
+
+  void addValidEntities(E entity) {
+    _validEntities.add(entity);
+  }
+
+  void maybeInclude(long windowIndex,
+                    MetricSampleCompleteness<G, E> completeness,
+                    AggregationOptions<G, E> options) {
+    Set<G> validGroupsForWindow = fillInCoverage(windowIndex, completeness, options);
+    Set<E> validEntitiesForWindow = 
+        options.granularity() == ENTITY ? _validEntities : validEntitiesWithGroupGranularity(validGroupsForWindow);
+    boolean meetEntityCoverageAfterMerge = meetEntityCoverageAfterMerge(completeness, validEntitiesForWindow, options);
+    boolean meetGroupCoverageAfterMerge = meetEntityGroupCoverageAfterMerge(completeness, validGroupsForWindow, options);
+
+    if (meetEntityCoverageAfterMerge && meetGroupCoverageAfterMerge) {
+      completeness.retainAllCoveredEntities(validEntitiesForWindow);
+      completeness.retainAllCoveredEntityGroups(validGroupsForWindow);
+      completeness.addValidWindowIndex(windowIndex);
+    }
+  }
+
+  private boolean meetEntityCoverageAfterMerge(MetricSampleCompleteness<G, E> completeness,
+                                               Set<E> validEntitiesForWindow,
+                                               AggregationOptions<G, E> options) {
+    int totalNumEntities = options.interestedEntities().size();
```

Comment:
Can the list of entities change?  It seems like the aggregator has it's own way of creating a list of entities.  Also how are new entities added to the system?